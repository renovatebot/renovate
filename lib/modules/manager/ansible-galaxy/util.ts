import { regEx } from '../../../util/regex';

export const newBlockRegEx = /^\s*-\s*((\w+):\s*(.*))$/;
export const blockLineRegEx = /^\s*((\w+):\s*(.*))$/;
export const galaxyDepRegex = /[\w-]+\.[\w-]+/;
export const dependencyRegex = /^dependencies:/;
export const galaxyRegEx = regEx(
  /^\s+(?<packageName>[\w.]+):\s*["'](?<version>.+)["']\s*/
);
export const nameMatchRegex = regEx(
  /(?<source>((git\+)?(?:(git|ssh|https?):\/\/)?(.*@)?(?<hostname>[\w.-]+)(?:(:\d+)?\/|:))(?<depName>[\w./-]+)(?:\.git)?)(,(?<version>[\w.]*))?/
);
