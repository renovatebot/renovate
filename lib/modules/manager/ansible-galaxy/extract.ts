import { logger } from '../../../logger';
import { newlineRegex, regEx } from '../../../util/regex';
import type { PackageDependency, PackageFileContent } from '../types';
import { extractCollections } from './collections';
import { extractCollectionsMetaDataFile } from './collections-metadata';
import { extractRoles } from './roles';

export function getSliceEndNumber(
  start: number,
  numberOfLines: number,
  ...blocks: number[]
): number {
  if (start < 0 || start > numberOfLines - 1) {
    return -1;
  }
  let nearestEnd = numberOfLines;
  for (const blocksKey of blocks) {
    if (start < blocksKey && blocksKey < nearestEnd) {
      nearestEnd = blocksKey;
    }
  }
  return nearestEnd;
}

export function extractPackageFile(
  content: string,
  packageFile: string
): PackageFileContent | null {
  logger.trace(`ansible-galaxy.extractPackageFile(${packageFile})`);
  const galaxyFileNameRegEx = regEx(/galaxy\.ya?ml$/);
  const deps: PackageDependency[] = [];
  const lines = content.split(newlineRegex);

  try {
    // if this is a galaxy.yml file we have to interpret the dependencies differently
    if (galaxyFileNameRegEx.exec(packageFile)) {
      const galaxyDeps = extractCollectionsMetaDataFile(lines);
      deps.push(...galaxyDeps);
    } else {
      // interpret requirements file
      // check if new or old format is used and save start lines for collection and roles.
      const positions = {
        collections: -1,
        roles: -1,
      };
      // find role and collection block
      lines.forEach((line, index) => {
        if (regEx(/^collections:/).exec(line)) {
          positions.collections = index;
        }
        if (regEx(/^roles:/).exec(line)) {
          positions.roles = index;
        }
      });
      if (positions.collections >= 0 || positions.roles >= 0) {
        // using new format
        const collectionLines = lines.slice(
          positions.collections,
          getSliceEndNumber(
            positions.collections,
            lines.length,
            positions.roles
          )
        );
        const collectionDeps = extractCollections(collectionLines);
        deps.push(...collectionDeps);

        const roleLines = lines.slice(
          positions.roles,
          getSliceEndNumber(
            positions.roles,
            lines.length,
            positions.collections
          )
        );
        const roleDeps = extractRoles(roleLines);
        deps.push(...roleDeps);
      } else {
        // use old format which only has only roles
        const galaxyDeps = extractRoles(lines);
        deps.push(...galaxyDeps);
      }
    }

    if (!deps.length) {
      return null;
    }
    return { deps };
  } catch (err) /* istanbul ignore next */ {
    logger.debug({ err, packageFile }, 'Error extracting ansible-galaxy deps');
    return null;
  }
}
