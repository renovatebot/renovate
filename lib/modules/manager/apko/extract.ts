import { logger } from '../../../logger';
import { getSiblingFileName, readLocalFile } from '../../../util/fs';
import { parseSingleYaml } from '../../../util/yaml';
import { ApkDatasource } from '../../datasource/apk';
import { id as apkVersioning } from '../../versioning/apk';
import type { PackageDependency, PackageFileContent } from '../types';

// Simple schema for apko configuration
interface ApkoConfig {
  contents?: {
    repositories?: string[];
    packages?: string[];
  };
  environment?: Record<string, string>;
  cmd?: string;
  archs?: string[];
}

// Schema for apko lock file
interface ApkoLockFile {
  version: string;
  config: {
    name: string;
    checksum: string;
  };
  contents: {
    keyring: string[];
    build_repositories: string[];
    runtime_repositories: string[];
    repositories: {
      name: string;
      url: string;
      architecture: string;
    }[];
    packages: {
      name: string;
      url: string;
      version: string;
      architecture: string;
      checksum: string;
    }[];
  };
}

// apko allows arch alias, see https://github.com/chainguard-dev/apko/blob/5244a17ae460bb053b7d13302d48fd0046aeb387/pkg/apk/apk/arch.go
function translateArch(arch: string): string {
  switch (arch) {
    case 'i386':
    case '386':
      return 'x86';
    case 'amd64':
      return 'x86_64';
    case 'arm64':
      return 'aarch64';
    case 'arm/v6':
      return 'armhf';
    case 'arm/v7':
      return 'armv7';
    default:
      return arch;
  }
}

export async function extractPackageFile(
  content: string,
  packageFile: string,
): Promise<PackageFileContent | null> {
  logger.trace('apko.extractPackageFile()');

  try {
    const parsed = parseSingleYaml<ApkoConfig>(content);
    const deps: PackageDependency[] = [];

    // Check if archs is specified
    if (!parsed.archs) {
      throw new Error('archs must be specified');
    }

    // Extract packages from the contents.packages array
    if (parsed.contents?.packages) {
      // determine the registry URLs based on the archs
      const registryUrls = parsed.archs.map((arch) => {
        const translatedArch = translateArch(arch);
        return `${parsed.contents?.repositories?.[0]}/${translatedArch}`;
      });

      for (const pkg of parsed.contents.packages) {
        // Try to extract version from package.
        // format is name{@tag}{[<>~=]version} - from https://wiki.alpinelinux.org/wiki/Alpine_Package_Keeper#World
        const versionMatch = /^(.+)[=><~^][=]?(.+)$/.exec(pkg);

        if (versionMatch) {
          const [, depName, currentValue] = versionMatch;
          deps.push({
            datasource: ApkDatasource.id,
            depName,
            currentValue,
            versioning: apkVersioning,
            registryUrls,
          });
        } else {
          // Package without version - add as unversioned
          deps.push({
            datasource: ApkDatasource.id,
            depName: pkg,
            skipReason: 'not-a-version',
            registryUrls,
          });
        }
      }
    }

    if (!deps.length) {
      return null;
    }

    // Derive lockfile name from package file name
    // If package file is 'image.yaml', lockfile should be 'image.lock.json'
    const packageFileName =
      packageFile.split('/').pop() ?? 'apko.yaml'; /* v8 ignore next */
    const baseName = packageFileName.replace(/\.ya?ml$/, '');
    const lockFileName = getSiblingFileName(
      packageFile,
      `${baseName}.lock.json`,
    );
    const lockFileContent = await readLocalFile(lockFileName, 'utf8');

    let lockFiles: string[] | undefined;

    if (lockFileContent) {
      try {
        const lockFile = parseSingleYaml<ApkoLockFile>(lockFileContent);

        // Create a mapping of package names to locked versions
        const lockedVersions = new Map<string, string>();
        const lockedPackages = new Set<string>();

        // Process all packages from the lock file
        for (const pkg of lockFile.contents?.packages || []) {
          lockedVersions.set(pkg.name, pkg.version);
          lockedPackages.add(pkg.name);
        }

        // Add locked versions to dependencies that are in both files
        for (const dep of deps) {
          if (dep.depName && lockedVersions.has(dep.depName)) {
            dep.lockedVersion = lockedVersions.get(dep.depName);
            // Keep architecture-specific URLs for locked packages
          }
        }

        logger.debug(
          { packageFile, lockFileName, packageCount: deps.length },
          'Found apko.lock.json with locked versions',
        );
        lockFiles = [lockFileName];
      } catch (err) {
        logger.debug({ err, lockFileName }, 'Error parsing apko.lock.json');
        // Use architecture-specific repository URLs when lock file parsing fails
        const registryUrls =
          parsed.archs?.map((arch) => {
            const translatedArch = translateArch(arch);
            return `${parsed.contents?.repositories?.[0]}/${translatedArch}`;
          }) || parsed.contents?.repositories; /* v8 ignore next */

        for (const dep of deps) {
          dep.registryUrls = registryUrls;
        }
        lockFiles = ['not here'];
      }
    }

    return {
      deps,
      lockFiles,
    };
  } catch (err) {
    logger.debug({ err, packageFile }, 'Error parsing apko YAML configuration');
    return null;
  }
}
