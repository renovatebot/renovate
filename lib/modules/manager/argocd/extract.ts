import is from '@sindresorhus/is';
import { loadAll } from 'js-yaml';
import { logger } from '../../../logger';
import { coerceArray } from '../../../util/array';
import { trimTrailingSlash } from '../../../util/url';
import { DockerDatasource } from '../../datasource/docker';
import { GitTagsDatasource } from '../../datasource/git-tags';
import { HelmDatasource } from '../../datasource/helm';
import type {
  ExtractConfig,
  PackageDependency,
  PackageFileContent,
} from '../types';
import type {
  ApplicationDefinition,
  ApplicationSource,
  ApplicationSpec,
} from './types';
import { fileTestRegex } from './util';

export function extractPackageFile(
  content: string,
  packageFile: string,
  _config?: ExtractConfig
): PackageFileContent | null {
  // check for argo reference. API version for the kind attribute is used
  if (fileTestRegex.test(content) === false) {
    return null;
  }

  let definitions: ApplicationDefinition[];
  try {
    definitions = loadAll(content) as ApplicationDefinition[];
  } catch (err) {
    logger.debug({ err, packageFile }, 'Failed to parse ArgoCD definition.');
    return null;
  }

  const deps = definitions.filter(is.plainObject).flatMap(processAppSpec);

  return deps.length ? { deps } : null;
}

function processSource(source: ApplicationSource): PackageDependency | null {
  if (
    !source ||
    !is.nonEmptyString(source.repoURL) ||
    !is.nonEmptyString(source.targetRevision)
  ) {
    return null;
  }

  // a chart variable is defined this is helm declaration
  if (source.chart) {
    // assume OCI helm chart if repoURL doesn't contain explicit protocol
    if (
      source.repoURL.startsWith('oci://') ||
      !source.repoURL.includes('://')
    ) {
      let registryURL = source.repoURL.replace('oci://', '');
      registryURL = trimTrailingSlash(registryURL);

      return {
        depName: `${registryURL}/${source.chart}`,
        currentValue: source.targetRevision,
        datasource: DockerDatasource.id,
      };
    }

    return {
      depName: source.chart,
      registryUrls: [source.repoURL],
      currentValue: source.targetRevision,
      datasource: HelmDatasource.id,
    };
  }

  return {
    depName: source.repoURL,
    currentValue: source.targetRevision,
    datasource: GitTagsDatasource.id,
  };
}

function processAppSpec(
  definition: ApplicationDefinition
): PackageDependency[] {
  const spec: ApplicationSpec | null | undefined =
    definition.kind === 'Application'
      ? definition?.spec
      : definition?.spec?.template?.spec;

  if (is.nullOrUndefined(spec)) {
    return [];
  }

  const deps: (PackageDependency | null)[] = [];

  if (is.nonEmptyObject(spec.source)) {
    deps.push(processSource(spec.source));
  }

  for (const source of coerceArray(spec.sources)) {
    deps.push(processSource(source));
  }

  return deps.filter(is.truthy);
}
