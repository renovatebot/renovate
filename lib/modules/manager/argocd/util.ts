import { regEx } from '../../../util/regex';

export const keyValueExtractionRegex = regEx(
  /^\s*(?<key>[^\s]+):\s+"?(?<value>[^"\s]+)"?\s*$/
);
// looks for `apiVersion: argoproj.io/
export const fileTestRegex = regEx(/\s*apiVersion:\s*argoproj.io\/\s*/);
