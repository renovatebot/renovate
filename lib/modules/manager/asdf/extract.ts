import is from '@sindresorhus/is';
import { logger } from '../../../logger';
import { isSkipComment } from '../../../util/ignore';
import { regEx } from '../../../util/regex';
import type { PackageDependency, PackageFileContent } from '../types';
import { upgradeableTooling } from './upgradeable-tooling';

export function extractPackageFile(content: string): PackageFileContent | null {
  logger.trace(`asdf.extractPackageFile()`);

  const regex = regEx(
    /^(?<toolName>([\w_-]+)) (?<version>[^\s#]+)(?: [^\s#]+)* *(?: #(?<comment>.*))?$/gm
  );

  const deps: PackageDependency[] = [];

  for (const groups of [...content.matchAll(regex)]
    .map((m) => m.groups)
    .filter(is.truthy)) {
    const depName = groups.toolName.trim();
    const version = groups.version.trim();

    const toolConfig = upgradeableTooling[depName];
    const toolDefinition = toolConfig
      ? typeof toolConfig.config === 'function'
        ? toolConfig.config(version)
        : toolConfig.config
      : undefined;

    if (toolDefinition) {
      const dep: PackageDependency = {
        currentValue: version,
        depName,
        ...toolDefinition,
      };
      if (isSkipComment((groups.comment ?? '').trim())) {
        dep.skipReason = 'ignored';
      }

      deps.push(dep);
    } else {
      const dep: PackageDependency = {
        depName,
        skipReason: 'unsupported-datasource',
      };

      deps.push(dep);
    }
  }

  return deps.length ? { deps } : null;
}
