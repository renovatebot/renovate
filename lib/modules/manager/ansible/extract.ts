import { logger } from '../../../logger';
import { newlineRegex, regEx } from '../../../util/regex';
import * as dockerVersioning from '../../versioning/docker';
import { getDep } from '../dockerfile/extract';
import type {
  ExtractConfig,
  PackageDependency,
  PackageFileContent,
} from '../types';

export function extractPackageFile(
  content: string,
  packageFile: string,
  config: ExtractConfig
): PackageFileContent | null {
  logger.trace(`ansible.extractPackageFile(${packageFile})`);
  let deps: PackageDependency[] = [];
  const re = regEx(/^\s*image:\s*'?"?([^\s'"]+)'?"?\s*$/);
  for (const line of content.split(newlineRegex)) {
    const match = re.exec(line);
    if (match) {
      const currentFrom = match[1];
      const dep = getDep(currentFrom, true, config.registryAliases);
      logger.debug(
        {
          depName: dep.depName,
          currentValue: dep.currentValue,
          currentDigest: dep.currentDigest,
        },
        'Docker image inside ansible'
      );
      dep.versioning = dockerVersioning.id;
      deps.push(dep);
    }
  }
  deps = deps.filter((dep) => !dep.currentValue?.includes('${'));
  if (!deps.length) {
    return null;
  }
  return { deps };
}
