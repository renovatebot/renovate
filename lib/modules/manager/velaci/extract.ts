import { load } from 'js-yaml';
import { logger } from '../../../logger';
import { coerceArray } from '../../../util/array';
import { getDep } from '../dockerfile/extract';
import type { PackageDependency, PackageFileContent } from '../types';
import type { VelaPipelineConfiguration } from './types';

export function extractPackageFile(
  file: string,
  packageFile?: string
): PackageFileContent | null {
  let doc: VelaPipelineConfiguration | undefined;

  try {
    doc = load(file, { json: true }) as VelaPipelineConfiguration;
  } catch (err) {
    logger.debug({ err, packageFile }, 'Failed to parse Vela file.');
    return null;
  }

  const deps: PackageDependency[] = [];

  // iterate over steps
  for (const step of coerceArray(doc.steps)) {
    const dep = getDep(step.image);

    deps.push(dep);
  }

  // iterate over services
  for (const service of coerceArray(doc.services)) {
    const dep = getDep(service.image);

    deps.push(dep);
  }

  // iterate over stages
  for (const stage of Object.values(doc.stages ?? {})) {
    for (const step of coerceArray(stage.steps)) {
      const dep = getDep(step.image);

      deps.push(dep);
    }
  }

  // check secrets
  for (const secret of Object.values(doc.secrets ?? {})) {
    if (secret.origin) {
      const dep = getDep(secret.origin.image);

      deps.push(dep);
    }
  }

  if (!deps.length) {
    return null;
  }

  return { deps };
}
