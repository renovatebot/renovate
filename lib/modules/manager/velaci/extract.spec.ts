import { loadFixture } from '../../../../test/util';
import { extractPackageFile } from './extract';

const invalidYAML = loadFixture('invalid.yml');
const stepsPipeline = loadFixture('.vela-steps.yml');
const servicesPipeline = loadFixture('.vela-services.yml');
const secretsPipeline = loadFixture('.vela-secrets.yml');
const stagesPipeline = loadFixture('.vela-stages.yaml');

describe('modules/manager/velaci/extract', () => {
  describe('extractPackageFile()', () => {
    it('should handle invalid YAML', () => {
      const res = extractPackageFile(invalidYAML);
      expect(res).toBeNull();
    });

    it('extracts multiple step pipeline image lines', () => {
      const res = extractPackageFile(stepsPipeline);
      expect(res.deps).toMatchSnapshot();
      expect(res.deps).toHaveLength(2);
    });

    it('extracts multiple services pipeline image lines', () => {
      const res = extractPackageFile(servicesPipeline);
      expect(res.deps).toMatchSnapshot();
      expect(res.deps).toHaveLength(3);
    });

    it('extracts multiple stages pipeline image lines', () => {
      const res = extractPackageFile(stagesPipeline);
      expect(res.deps).toMatchSnapshot();
      expect(res.deps).toHaveLength(2);
    });

    it('extracts multiple secrets pipeline image lines', () => {
      const res = extractPackageFile(secretsPipeline);
      expect(res.deps).toMatchSnapshot();
      expect(res.deps).toHaveLength(2);
    });
  });
});
