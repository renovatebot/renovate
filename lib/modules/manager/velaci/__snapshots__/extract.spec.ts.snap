// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`modules/manager/velaci/extract extractPackageFile() extracts multiple secrets pipeline image lines 1`] = `
Array [
  Object {
    "autoReplaceStringTemplate": "{{packageName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}",
    "currentDigest": "sha256:5082d4db78ee2d24f72b25eb75537f2e19c49f04a595378d1ae8814d6f2fbf28",
    "currentValue": "10.0.0",
    "datasource": "docker",
    "depName": "node",
    "depType": "docker",
    "packageName": "amd64/node",
    "replaceString": "amd64/node:10.0.0@sha256:5082d4db78ee2d24f72b25eb75537f2e19c49f04a595378d1ae8814d6f2fbf28",
  },
  Object {
    "autoReplaceStringTemplate": "{{depName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}",
    "currentDigest": undefined,
    "currentValue": "v0.1.0",
    "datasource": "docker",
    "depName": "target/secret-vault",
    "depType": "docker",
    "replaceString": "target/secret-vault:v0.1.0",
  },
]
`;

exports[`modules/manager/velaci/extract extractPackageFile() extracts multiple services pipeline image lines 1`] = `
Array [
  Object {
    "autoReplaceStringTemplate": "{{packageName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}",
    "currentDigest": "sha256:5082d4db78ee2d24f72b25eb75537f2e19c49f04a595378d1ae8814d6f2fbf28",
    "currentValue": "10.0.0",
    "datasource": "docker",
    "depName": "node",
    "depType": "docker",
    "packageName": "amd64/node",
    "replaceString": "amd64/node:10.0.0@sha256:5082d4db78ee2d24f72b25eb75537f2e19c49f04a595378d1ae8814d6f2fbf28",
  },
  Object {
    "autoReplaceStringTemplate": "{{depName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}",
    "currentDigest": undefined,
    "currentValue": "5.7.24",
    "datasource": "docker",
    "depName": "mysql",
    "depType": "docker",
    "replaceString": "mysql:5.7.24",
  },
  Object {
    "autoReplaceStringTemplate": "{{depName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}",
    "currentDigest": undefined,
    "currentValue": "alpine",
    "datasource": "docker",
    "depName": "redis",
    "depType": "docker",
    "replaceString": "redis:alpine",
  },
]
`;

exports[`modules/manager/velaci/extract extractPackageFile() extracts multiple stages pipeline image lines 1`] = `
Array [
  Object {
    "autoReplaceStringTemplate": "{{depName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}",
    "currentDigest": undefined,
    "currentValue": "1.13",
    "datasource": "docker",
    "depName": "golang",
    "depType": "docker",
    "replaceString": "golang:1.13",
  },
  Object {
    "autoReplaceStringTemplate": "{{packageName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}",
    "currentDigest": "sha256:5082d4db78ee2d24f72b25eb75537f2e19c49f04a595378d1ae8814d6f2fbf28",
    "currentValue": "10.0.0",
    "datasource": "docker",
    "depName": "node",
    "depType": "docker",
    "packageName": "amd64/node",
    "replaceString": "amd64/node:10.0.0@sha256:5082d4db78ee2d24f72b25eb75537f2e19c49f04a595378d1ae8814d6f2fbf28",
  },
]
`;

exports[`modules/manager/velaci/extract extractPackageFile() extracts multiple step pipeline image lines 1`] = `
Array [
  Object {
    "autoReplaceStringTemplate": "{{depName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}",
    "currentDigest": undefined,
    "currentValue": "1.13",
    "datasource": "docker",
    "depName": "golang",
    "depType": "docker",
    "replaceString": "golang:1.13",
  },
  Object {
    "autoReplaceStringTemplate": "{{packageName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}",
    "currentDigest": "sha256:5082d4db78ee2d24f72b25eb75537f2e19c49f04a595378d1ae8814d6f2fbf28",
    "currentValue": "10.0.0",
    "datasource": "docker",
    "depName": "node",
    "depType": "docker",
    "packageName": "amd64/node",
    "replaceString": "amd64/node:10.0.0@sha256:5082d4db78ee2d24f72b25eb75537f2e19c49f04a595378d1ae8814d6f2fbf28",
  },
]
`;
